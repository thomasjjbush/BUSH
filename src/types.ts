import { ReactElement, UIEvent } from 'react';

export interface AspectRatioProps {
    children?: ReactElement | ReactElement[] | string;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '3:2' = 66.66,
    '4:3' = 75,
    '8:5' = 62.5,
    '16:9' = 56.25,
    '32:9' = 28.125,
}

export interface ButtonProps {
    ariaLabel?: string;
    children: Children;
    design?: Buttons;
    onClick: () => void;
}

export enum Buttons {
    PRIMARY,
    SECONDARY,
}

export interface CarouselProps {
    children: ReactElement[];
    color?: string;
    itemsToDisplay?: number;
    margin?: number;
    numberOfItems: number;
    startingIndex?: number;
}

type Children = ReactElement | ReactElement[] | string | string[] | (ReactElement | string)[];

export interface Client {
    logo: {
        url: string;
    };
    name: string;
    primaryColor: string;
    secondaryColor?: string;
    slug: string;
    url: string;
}

export enum ClientsActions {
    LOADED = 'clients/LOADED',
}

export interface ClientProps {
    alt?: string;
    background?: boolean;
    border?: string;
    logo: string;
    size: number;
    url?: string;
    width: number;
}

type Collection<T> = { items: T[]; total?: number };

export interface Employment {
    companyName: string;
    endDate?: string;
    location: {
        lat: string;
        lon: string;
    };
    responsibilities: string;
    startDate: string;
    title: string;
    url: string;
}

export enum EmploymentActions {
    LOADED = 'employment/LOADED',
}

export interface ErrorProps {
    button?: ReactElement;
    code: number;
}

export interface Errors {
    [key: number]: string;
}

type Gallery<T> = Collection<T> & { isLandScape: boolean };

export type GetState = () => Store;

export interface HiddenTextProps {
    children: string;
    lines: number;
}

export interface IconProps {
    color?: string;
    disabled?: boolean;
    icon: Icons | string;
    onClick?: () => void;
    size?: number;
    testID?: string;
}

export enum Icons {
    CONTRAST = 'e9d5',
    ERROR = 'ea07',
    GITHUB = 'eab0',
    INFO = 'ea0c',
    LEFT = 'ea44',
    LINKEDIN = 'eac9',
    RIGHT = 'ea42',
    WHATSAPP = 'ea93',
}

export interface ImageProps {
    alt?: string;
    height?: number;
    quality?: number;
    size: number;
    src: string;
    width?: number | string;
}

interface Media {
    description?: string;
    name: string;
    url: string;
}

export enum Platform {
    DESKTOP,
    MOBILE,
    TABLET,
}

export interface Project {
    client: Client;
    description: string;
    gallery?: Gallery<Media>;
    hero: Media;
    name: string;
    primaryTag?: Tag;
    responsibilities: Collection<Responsibility>;
    slug: string;
    tags: Collection<Tag>;
    thumbnail: Media;
    video: Media;
    year: number;
}

export enum ProjectActions {
    CLEARED = 'project/CLEARED',
    ERROR = 'project/ERROR',
    LOADED_PROJECT = 'project/LOADED',
}

export interface ProjectState {
    error?: number;
    item?: Project;
    sameClient?: Partial<Project>[];
    sameTag?: Partial<Project>[];
}

export enum ProjectsActions {
    LOADED = 'projects/LOADED',
    LOADED_MORE = 'projects/LOADED_MORE',
    LOADING_MORE = 'projects/LOADING_MORE',
}

export interface ProjectsState extends Collection<Project> {
    loading?: boolean;
}

export interface Responsibility {
    description: string;
    icon: string;
    name: string;
}

export interface ScrollerProps {
    children: Children;
    loading?: boolean;
    loadMore?: () => void;
    numberOfItems?: number;
    onScroll?: (e: UIEvent<HTMLDivElement>) => void;
    testID?: string;
    totalItems?: number;
}

export interface Store {
    clients: Client[];
    employment: Employment[];
    project: ProjectState;
    projects: ProjectsState;
    tags: Tag[];
}

export type StyledProps<T = {}> = { theme: Theme } & T;

export interface Tag {
    name: string;
    slug: string;
}

export enum TagActions {
    LOADED = 'tag/LOADED',
}

export interface TagProps {
    tags: Tag[];
}

export interface Theme {
    breakpoints: {
        mobile: number;
        tabletLandscape: number;
        tabletPortrait: number;
    };
    colors: {
        highlight: string;
        background: string;
        border: string;
        brand: string;
        offset: string;
        text: string;
    };
    images: {
        logo: {
            pink: string;
            white: string;
        };
        loading: string;
    };
    isWebpSupported?: boolean;
}

export interface ThumbnailProps {
    client?: ReactElement;
    color: string;
    name: string;
    ratio?: AspectRatios;
    url: string;
}

declare global {
    interface Window extends Window {
        __REDUX_DEVTOOLS_EXTENSION__?: Function;
        __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: Function;
    }
}
