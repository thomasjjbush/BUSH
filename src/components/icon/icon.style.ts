import styled from 'styled-components';
import { IconProps as Props } from '../../types';
import { StyledButton } from '../button/button.style';

const shared = (color: string, icon: string, size: number): string => `
	display: flex;
	font-family: 'icomoon' !important;
	speak: never;
	font-style: normal;
	font-weight: normal;
	font-variant: normal;
	text-transform: none;
	line-height: 1;
	-webkit-font-smoothing: antialiased;
	-moz-osx-font-smoothing: grayscale;
	color: ${color};
	font-size: ${size || 20}px;

	:after {
		content: "\\${icon}";
		flex: 1;
		display: flex;
		justify-content: center;
		align-items: center;
  	}
`;

export const ClickableIcon = styled(StyledButton)<Props>`
    ${(props): string => shared(props.color || props.theme.colors.brand, props.icon, props.size)}
    align-self: auto;
    appearance: none;
    background: transparent;
    border: 0;
    margin: 0;
    padding: 0;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.4;
    }
`;

export const Icon = styled.i<Props>`
    ${(props): string => shared(props.color || props.theme.colors.brand, props.icon, props.size)}
`;
