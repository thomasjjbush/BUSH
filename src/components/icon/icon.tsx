import React, { FunctionComponent, ReactElement } from 'react';
import { IconProps as Props } from '../../types';
import { StyledIcon } from './icon.style';

export const Icon: FunctionComponent<Props> = (props: Props): ReactElement => (
    <StyledIcon data-test-id={props.testID || 'icon'} {...props} {...(props.onClick && { tabIndex: 0 })} />
);

Icon.displayName = 'Icon';
