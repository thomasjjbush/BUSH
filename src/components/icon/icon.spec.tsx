import React from 'react';
import { shallow } from 'enzyme';
import { Icon } from './icon';
import { Icons } from '../../types';
import * as Styled from './icon.style';

describe('Icon', () => {
    it('should match snapshot', () => {
        expect(shallow(<Icon icon={Icons.CONTRAST} />)).toMatchSnapshot();
    });

    it('should match snapshot when optional args are provided', () => {
        expect(shallow(<Icon color="red" disabled icon={Icons.CONTRAST} size={60} />)).toMatchSnapshot();
    });

    it('should invoke onClick cb on click and set tab-index to 0 if onClick is provided', () => {
        const onClick = jest.fn();
        const wrapper = shallow(<Icon icon={Icons.CONTRAST} onClick={onClick} />);
        wrapper.find(Styled.ClickableIcon).simulate('click');

        expect(onClick).toHaveBeenCalledTimes(1);
    });
});
