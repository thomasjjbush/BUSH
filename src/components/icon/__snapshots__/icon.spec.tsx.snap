// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should match snapshot 1`] = `
<styled.i
  data-test-id="icon"
  icon="e9d5"
  styles="
    
	display: flex;
	font-family: 'icomoon' !important;
	speak: never;
	font-style: normal;
	font-weight: normal;
	font-variant: normal;
	text-transform: none;
	line-height: 1;
	-webkit-font-smoothing: antialiased;
	-moz-osx-font-smoothing: grayscale;
	color: theme.colors.brand;
	font-size: 20px;

	:after {
		content: \\"\\\\e9d5\\";
		flex: 1;
		display: flex;
		justify-content: center;
		align-items: center;
  	}

"
/>
`;

exports[`Icon should match snapshot when optional args are provided 1`] = `
<styled.i
  color="red"
  data-test-id="icon"
  disabled={true}
  icon="e9d5"
  size={60}
  styles="
    
	display: flex;
	font-family: 'icomoon' !important;
	speak: never;
	font-style: normal;
	font-weight: normal;
	font-variant: normal;
	text-transform: none;
	line-height: 1;
	-webkit-font-smoothing: antialiased;
	-moz-osx-font-smoothing: grayscale;
	color: red;
	font-size: 60px;

	:after {
		content: \\"\\\\e9d5\\";
		flex: 1;
		display: flex;
		justify-content: center;
		align-items: center;
  	}

"
/>
`;
